File: 473. Matchsticks to Square.cpp
class Solution {
public:
    bool backtrack(long a, long b, long c, long d, const std::vector<int>& matchsticks, int index, long target, std::unordered_map<std::string, bool>& hm) {
        if (index >= (int)matchsticks.size()) {
            return (a == b && b == c && c == d);
        }
        std::string key = std::to_string(a) + "-" + std::to_string(b) + "-" + std::to_string(c) + "-" + std::to_string(d) + "-" + std::to_string(index);
        if (hm.find(key) != hm.end()) return hm[key];

        int curr = matchsticks[index];
        bool res = false;
        for (int i = 0; i < 4; i++) {
            if (i == 0 && curr + a <= target) res |= backtrack(a + curr, b, c, d, matchsticks, index + 1, target, hm);
            else if (i == 1 && curr + b <= target) res |= backtrack(a, b + curr, c, d, matchsticks, index + 1, target, hm);
            else if (i == 2 && curr + c <= target) res |= backtrack(a, b, c + curr, d, matchsticks, index + 1, target, hm);
            else if (i == 3 && curr + d <= target) res |= backtrack(a, b, c, d + curr, matchsticks, index + 1, target, hm);
        }

        hm[key] = res;
        return res;
    }

    bool makesquare(std::vector<int>& ms) {
        std::sort(ms.begin(), ms.end());
        int n = (int)ms.size();
        for (int i = 0; i < n / 2; i++) {
            std::swap(ms[i], ms[n - i - 1]);
        }
        long target = 0;
        for (int stick : ms) target += stick;

        if (target % 4 != 0) return false;
        std::unordered_map<std::string, bool> hm;
        return backtrack(0, 0, 0, 0, ms, 0, target / 4, hm);
    }
};

File: 473. Matchsticks to Square.java
class Solution {
    public boolean dfs(int[] matches,int index, long[] buckets, long sum){
        if(index >= matches.length){
            if(buckets[0] == buckets[1] && buckets[1] == buckets[2] && buckets[2] == buckets[3] && buckets[0] == sum) return true;
            return false;
        }

        if(buckets[0] + matches[index] <= sum){
            buckets[0] += matches[index];
            boolean l0 = dfs(matches, index + 1, buckets, sum);
            buckets[0] -= matches[index];
            if(l0) return true;
        }

        if(buckets[1] + matches[index] <= sum){
            buckets[1] += matches[index];
            boolean l1 = dfs(matches, index + 1, buckets, sum);
            buckets[1] -= matches[index];
            if(l1) return true;
        }

        if(buckets[2] + matches[index] <= sum){
            buckets[2] += matches[index];
            boolean l3 = dfs(matches, index + 1, buckets, sum);
            buckets[2] -= matches[index];
            if(l3) return true;
        }

        if(buckets[3] + matches[index] <= sum){
            buckets[3] += matches[index];
            boolean l4 = dfs(matches, index + 1, buckets, sum);
            buckets[3] -= matches[index];
            if(l4) return true;
        }

        return false;
    }
    public boolean makesquare(int[] matchsticks) {
        Arrays.sort(matchsticks);
        for(int i = 0; i < matchsticks.length / 2; i++){
            int temp = matchsticks[i];
            matchsticks[i] = matchsticks[matchsticks.length - i - 1];
            matchsticks[matchsticks.length - i - 1] = temp;
        }
        long sum = 0;
        for(int matches : matchsticks) sum += matches;
        if(sum % 4 != 0) return false;
        sum /= 4;

        return dfs(matchsticks, 0, new long[]{0, 0, 0, 0}, sum);
    }
}